lib/types.js

```diff
@@ -2142,36 +2147,42 @@ RecordType.prototype._createConstructor = function (errorStackTraces) {
       // particular, without a default) to populate a stack trace below.
       stackField = field;
     }
-    innerArgs.push('v' + i);
-    innerBody += '  ';
-    if (!hasDefault) {
-      innerBody += 'this.' + name + ' = v' + i + ';\n';
-    } else {
-      innerBody += 'if (v' + i + ' === undefined) { ';
-      innerBody += 'this.' + name + ' = d' + ds.length + '(); ';
-      innerBody += '} else { this.' + name + ' = v' + i + '; }\n';
-      outerArgs.push('d' + ds.length);
-      ds.push(defaultValue);
-    }
-  }
-  if (stackField) {
-    // We should populate a stack trace.
-    innerBody += '  if (this.stack === undefined) { ';
-    /* istanbul ignore else */
-    if (typeof Error.captureStackTrace == 'function') {
-      // v8 runtimes, the easy case.
-      innerBody += 'Error.captureStackTrace(this, this.constructor);';
-    } else {
-      // A few other runtimes (e.g. SpiderMonkey), might not work everywhere.
-      innerBody += 'this.stack = Error().stack;';
+    if (hasDefault) {
+      ds[i] = defaultValue;
     }
-    innerBody += ' }\n';
   }
-  var outerBody = 'return function ' + this._getConstructorName() + '(';
-  outerBody += innerArgs.join() + ') {\n' + innerBody + '};';
-  var Record = new Function(outerArgs.join(), outerBody).apply(undefined, ds);
 
   var self = this;
+  function ConstructorFunction(outerArgs) {
+    const constructorName = self._getConstructorName();
+    var innerFunction = ({
+      [constructorName]: function(...innerArgs) {
```

Comment:
This is so that the function can get a dynamic name. This is again only available in ES2015. An example of it would be something like this,

```javascript
var fun1 = function() { console.log('fun1'); }
var funname = 'fun2';
var dfun = ({
  [funname]: function() { console.log('dynamic function'); }
})[funname];

dfuninstance = new dfun()
dynamic function
dfun {}
dfun.name
"fun2"
```

The aim was to have the constructor function with proper name when being invoked.

Reply:
Interesting, pretty cool. Let's extract the logic to a function and document it to help future readers.